File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
from collections import deque

def solution(s):
    string = deque(list(s))
    ans = 0
    for _ in range(len(s)):
        string.rotate(-1)
        stack = []
        for char in string:
            if not stack or char == '[' or char == '(' or char == '{':
                stack.append(char)
                continue

            if char == ']' and stack[-1] == '[':
                stack.pop()
            elif char == ')' and stack[-1] == '(':
                stack.pop()
            elif char == '}' and stack[-1] == '{':
                stack.pop()

        if not stack:
            ans += 1

    return ans

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;
import java.io.*;

class Solution {
    public int solution(String s) {
        HashMap<Character, Character> map = new HashMap<>(){{
            put(']', '[');
            put(')', '(');
            put('}', '{');
        }};

        ArrayDeque<Character> queue = new ArrayDeque<>();
        for (int i = 0; i < s.length(); i++) {
            queue.addLast(s.charAt(i));
        }

        int answer = 0;
        for (int i = 0; i < s.length(); i++) {
            queue.addLast(queue.pollFirst());
            Stack<Character> st = new Stack<>();
            for (Character ch : queue) {
                if (st.isEmpty()) {
                    st.push(ch);
                    continue;
                }

                if (map.containsKey(ch) && st.peek() == map.get(ch)) {
                    st.pop();
                } else {
                    st.push(ch);
                }
            }

            if (st.isEmpty()) {
                answer++;
            }
        }
        return answer;
    }
}